File: GitToolBox/src/main/java/zielu/gittoolbox/formatter/RegExpFormatter.java
package zielu.gittoolbox.formatter;

import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;
import jodd.util.StringBand;
import org.apache.commons.lang.StringUtils;

public class RegExpFormatter implements Formatter {
  private final Pattern pattern;

  private RegExpFormatter(Pattern pattern) {
    this.pattern = pattern;
  }

  public static RegExpFormatter create(String pattern) {
    if (StringUtils.isNotBlank(pattern)) {
      try {
        return new RegExpFormatter(Pattern.compile(pattern));
      } catch (PatternSyntaxException e) {
        return new RegExpFormatter(null);
      }
    } else {
      return new RegExpFormatter(null);
    }
  }

  @Override
  public Formatted format(String input) {
    if (pattern == null) {
      return new Formatted(input, false);
    } else {
      Matcher matcher = pattern.matcher(input);
      if (matcher.matches()) {
        return new Formatted(format(matcher), true);
      } else {
        return new Formatted(input, false);
      }
    }
  }

  private String format(Matcher matcher) {
    int count = matcher.groupCount();
    StringBand result = new StringBand(count);
    for (int i = 1; i <= count; i++) {
      result.append(matcher.group(i));
    }
    return result.toString();
  }
}

File: src/main/kotlin/zielu/gittoolbox/formatter/RegExpFormatter.kt
package zielu.gittoolbox.formatter

import jodd.util.StringBand
import zielu.gittoolbox.completion.FormatterIcons
import java.util.regex.Matcher
import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException
import javax.swing.Icon

internal class RegExpFormatter(
  private val pattern: Pattern?
) : Formatter {

  override fun format(input: String): Formatted {
    return if (pattern == null) {
      Formatted(input, false, "no pattern")
    } else {
      val matcher = pattern.matcher(input)
      return if (matcher.matches()) {
        Formatted(format(matcher), true, null)
      } else {
        Formatted(input, false, "pattern: ${pattern.pattern()}")
      }
    }
  }

  private fun format(matcher: Matcher): String {
    val count = matcher.groupCount()
    val result = StringBand(count)
    for (i in 1..count) {
      result.append(matcher.group(i))
    }
    return result.toString()
  }

  override fun getIcon(): Icon = FormatterIcons.RegExp

  companion object {
    private val empty = RegExpFormatter(null)

    @JvmStatic
    fun create(pattern: String?): Formatter {
      return if (pattern != null && pattern.isNotBlank()) {
        try {
          RegExpFormatter(Pattern.compile(pattern))
        } catch (e: PatternSyntaxException) {
          empty
        }
      } else {
        empty
      }
    }
  }
}
